File: advall-python/d15.py
# slot model: (int noOfPositions, int startPosition)

def openWhenReached(slotNum, slot, timeOfPush):
    posWhenReached = (slot[1] + slotNum + timeOfPush) % slot[0]
    return posWhenReached == 0

def allOpen(slots, timeOfPush):
    opens = []
    for i, slot in enumerate(slots):
        opens.append(openWhenReached((i+1), slot, timeOfPush))
    return all(opens)

def whenToPush(slots):
    i = 0
    while True:
        if allOpen(slots, i):
            return i
        i += 1

def parseInput(lines):
    slots = []
    for line in lines:
        words = line.split()
        noOfPos = int(words[3])
        atPos = int(words[-1][0:-1])
        slot = (noOfPos, atPos)
        slots.append(slot)
    return slots

with open ("d15_input.txt", "r") as inputfile:
    input = inputfile.readlines()

slots = parseInput(input)
print "Part one: " + str(whenToPush(slots))
slots.append((11,0))
print "Part two: " + str(whenToPush(slots))

File: advall-haskell/d15.hs
import Data.List

parseInput :: String -> [(Integer, Integer)]
parseInput s = map wordsToSlots (map words (lines s))

wordsToSlots :: [String] -> (Integer, Integer)
wordsToSlots s = (read (s!!3), read (init (s!!11)))

allOpen :: [(Integer, Integer)] -> Integer -> Bool
allOpen slots timeOfPush = and (map (openWhenReached timeOfPush) (zip [1..] slots))

openWhenReached :: Integer -> (Integer, (Integer, Integer)) -> Bool
openWhenReached timeOfPush (offset, (noOfPos, startPos)) = ((startPos + offset + timeOfPush) `mod` noOfPos) == 0

solve :: [(Integer, Integer)] -> Integer -> Integer
solve slots t
    | allOpen slots t = t
    | otherwise = solve slots (t+1)


main = do
    input <- readFile "d15_input.txt"
    let slots = parseInput input
    let answer1 = solve slots 0
    putStrLn $ "Part one: " ++ show answer1
    let slots2 = slots ++ [(11,0)]
    let answer2 = solve slots2 0
    putStrLn $ "Part two: " ++ show answer2

